In [1]:
import numpy as np

# 1.

Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks): zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]. Используя математические операции, посчитать коэффициенты линейной регрессии, приняв за X заработную плату (то есть, zp - признак), а за y - значения скорингового балла (то есть, ks - целевая переменная). Произвести расчет как с использованием intercept, так и без.

In [2]:
#X = np.array([1, 5, 10, 15, 20, 25, 30, 35])
X = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
X

array([ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110])

In [3]:
#y = np.array([[1, 1, 2, 5, 7, 15, 70, 100]])
y = np.array([[401, 574, 874, 919, 459, 739, 653, 902, 746, 832]])
y

array([[401, 574, 874, 919, 459, 739, 653, 902, 746, 832]])

In [4]:
if len(X) != len(y):
    n = len(X)
n

10

In [5]:
b = ((n*((X*y).sum())) - ((X.sum())*(y.sum()))) / ((n*((X**2).sum()))-((X.sum())**2))
b

2.6205388824027653

In [6]:
a = y.mean() - b * X.mean()
a

444.17735732435955

#### Модель парной линейной регрессии с "intercept"

y_hat = 444.18 + 2.62*X

Теперь построим модель парной линейной регресии без "intercept".

In [7]:
b = ((X*y).sum()) / ((X**2).sum())
b

5.889820420132689

#### Модель парной линейной регрессии без "intercept"

y_hat = 5.89*X

# 2.

Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept).

In [8]:
B1 = 0.1

In [9]:
alpha = 1e-6

In [10]:
def mse_(B1, y=y, X=X, n=n):
    return np.sum(((B1*X - y)**2)/n)

In [11]:
mse_(b)

56516.85841571943

In [12]:
for i in range(100000):
    B1 -= alpha*(2/n)*np.sum((B1*X-y)*X)
    if i%10000==0:    #показ заданной строки; в данном случае каждой десятитысячной
        print('Iteration: {i}, B1={B1}, mse={mse}'.format(i=i, B1=B1, mse=mse_(B1)))

Iteration: 0, B1=0.25952808, mse=493237.7212546963
Iteration: 10000, B1=5.889820420132673, mse=56516.85841571942
Iteration: 20000, B1=5.889820420132673, mse=56516.85841571942
Iteration: 30000, B1=5.889820420132673, mse=56516.85841571942
Iteration: 40000, B1=5.889820420132673, mse=56516.85841571942
Iteration: 50000, B1=5.889820420132673, mse=56516.85841571942
Iteration: 60000, B1=5.889820420132673, mse=56516.85841571942
Iteration: 70000, B1=5.889820420132673, mse=56516.85841571942
Iteration: 80000, B1=5.889820420132673, mse=56516.85841571942
Iteration: 90000, B1=5.889820420132673, mse=56516.85841571942


# 3.

В каких случаях для вычисления доверительных интервалов и проверки статистических гипотез используется таблица значений функции Лапласа, а в каких - таблица критических точек распределения Стьюдента?

1) При известном значении дисперсии генеральной совокупности используется таблица значений функции Лапласа;

2) При неизвестном значении дисперсии генеральной совокупности данная метрика расчитывается и ииспользуется таблица критических точек распределения Стьюдента.

# 4.

Произвести вычисления как в пункте 2, но с вычислением intercept. Учесть, что изменение коэффициентов должно производиться на каждом шаге одновременно (то есть изменение одного коэффициента не должно влиять на изменение другого во время одной итерации).

In [13]:
A2 = 100

In [14]:
B2 = 0.5

In [15]:
alpha2 = 1e-5

In [16]:
def mse_2(A2, B2, y=y, X=X, n=n):
    return np.sum((((A2 + B2*X) - y)**2)/n)

In [17]:
mse_2(444.17735732435955, 2.6205388824027653)

6470.414201176658

In [18]:
for i in range(1000000):
    A2 -= alpha2*(2/n)*np.sum(A2 + B2*X-y)
    B2 -= alpha2*(2/n)*np.sum((A2 + B2*X-y)*X)
    if i%100000==0:
        print('Iteration: {i}, A2 = {A2}, B2={B2}, mse={mse}'.format(i=i, A2=A2, B2=B2, mse=mse_2(A2, B2)))

Iteration: 0, A2 = 100.011184, B2=1.7822453188480003, mse=193112.37518353554
Iteration: 100000, A2 = 236.9655895316027, B2=4.145701370022048, mse=17361.95259005893
Iteration: 200000, A2 = 319.41258947794705, B2=3.538858076929363, mse=10419.024923573004
Iteration: 300000, A2 = 369.0549477161389, B2=3.1734702257066836, mse=7901.940775323632
Iteration: 400000, A2 = 398.9452255073789, B2=2.953465681962269, mse=6989.398845792202
Iteration: 500000, A2 = 416.9425313998575, B2=2.8209982255151482, mse=6658.566531850161
Iteration: 600000, A2 = 427.77893183914125, B2=2.741237931846179, mse=6538.626815760144
Iteration: 700000, A2 = 434.30366255063666, B2=2.693213272682308, mse=6495.143954837706
Iteration: 800000, A2 = 438.23228359929334, B2=2.664297031508808, mse=6479.379708801164
Iteration: 900000, A2 = 440.5977549154871, B2=2.6468862051091877, mse=6473.664550050602
